hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCleanerChore.java

```diff
@@ -0,0 +1,289 @@
+      // this will be skipped and won't be archived.
+      long maxCreationTimeToArchive = EnvironmentEdgeManager.currentTime() - minAgeToArchive;
+      LOG.info("Only MOB files whose creation time less than {} will be archived",
+        maxCreationTimeToArchive);
+      Path rootDir = FSUtils.getRootDir(conf);
+      Path tableDir = FSUtils.getTableDir(rootDir, table);
+      // How safe is this call?
+      List<Path> regionDirs = FSUtils.getRegionDirs(FileSystem.get(conf), tableDir);
+
+      Set<String> allActiveMobFileName = new HashSet<String>();
+      FileSystem fs = FileSystem.get(conf);
+      for (Path regionPath : regionDirs) {
+        for (ColumnFamilyDescriptor hcd : list) {
+          String family = hcd.getNameAsString();
+          Path storePath = new Path(regionPath, family);
+          boolean succeed = false;
+          Set<String> regionMobs = new HashSet<String>();
+          while (!succeed) {
+
+            RemoteIterator<LocatedFileStatus> rit = fs.listLocatedStatus(storePath);
+            List<Path> storeFiles = new ArrayList<Path>();
+            // Load list of store files first
+            while (rit.hasNext()) {
+              Path p = rit.next().getPath();
+              if (fs.isFile(p)) {
+                storeFiles.add(p);
+              }
+            }
+            try {
+              for (Path pp : storeFiles) {
+                LOG.debug("Store file: {}", pp);
+                HStoreFile sf =
+                    new HStoreFile(fs, pp, conf, CacheConfig.DISABLED, BloomType.NONE, true);
+                sf.initReader();
+                byte[] mobRefData = sf.getMetadataValue(HStoreFile.MOB_FILE_REFS);
+                byte[] bulkloadMarkerData = sf.getMetadataValue(HStoreFile.BULKLOAD_TASK_KEY);
+                // close store file to avoid memory leaks
+                sf.closeStoreFile(true);
+                if (mobRefData == null && bulkloadMarkerData == null) {
+                  LOG.warn("Found old store file with no MOB_FILE_REFS: {} - "
+                    + "can not proceed until all old files will be MOB-compacted.", pp);
+                  return;
+                } else if (mobRefData == null && bulkloadMarkerData != null) {
+                  LOG.info("Skipping file without MOB references (bulkloaded file):{}", pp);
+                  continue;
+                }
+                if (mobRefData.length > 1) {
+                  // if length = 1 means NULL, that there are no MOB references
+                  // in this store file, but the file was created by new MOB code
+                  String[] mobs = new String(mobRefData).split(",");
+                  LOG.debug("Found: {} mob refs: ", mobs.length, Arrays.toString(mobs));
+                  regionMobs.addAll(Arrays.asList(mobs));
+                }
+              }
+            } catch (FileNotFoundException e) {
+              LOG.warn("Starting MOB cleaning cycle from the beginning due to error:",e);
+              continue;
```

Comment:
should we clear `regionMobs` in this case? I guess not clearing it means we err on the side of keeping some files a bit longer, which isn't so bad.

Reply:
yes, we can keep them.